**Assignment 4**

In this assignment, I use greedy algorithm to create my own TSP solution that finds the shortest distance.


Hemingway came to lots of places in his life and he only lived in Oak Park, Paris, Milano,Toronto,Chicago,Key west, Havana,Vence, Entebbe,New York and Ketchum. 

#Import necessary package
 

In [46]:
from random import choices

#Cities where Hemingway lived 

In [47]:
city_list = ['Oak Park（Illinois）','Paris','Milano','Toronto','Chicago','Key West','Havana',
           'Venice','Entebbe','New York','Ketchum']
print (city_list)

['Oak Park（Illinois）', 'Paris', 'Milano', 'Toronto', 'Chicago', 'Key West', 'Havana', 'Venice', 'Entebbe', 'New York', 'Ketchum']


#Define a dictionary to store the distance between every two of these city.

In [48]:
distance_dic={}
distance_dic[6667] = {'Oak Park（Illinois）','Paris'}
distance_dic[7307] = {'Milano','Oak Park（Illinois）'}
distance_dic[712] = {'Oak Park（Illinois）','Toronto'}
distance_dic [12] ={'Oak Park（Illinois）','Chicago'}
distance_dic[2007] = {'Oak Park（Illinois）','Key West'}
distance_dic[2147] = {'Oak Park（Illinois）','Havana'}
distance_dic[7505] = {'Oak Park（Illinois）','Venice'}
distance_dic[12464] = {'Oak Park（Illinois）','Entebbe'}
distance_dic[1158] = {'Oak Park（Illinois）','New York'}
distance_dic[2171] = {'Oak Park（Illinois）','Ketchum'}

distance_dic [640] ={'Paris','Milano'}
distance_dic[6006] = {'Paris','Toronto'}
distance_dic[6661] = {'Paris','Chicago'}
distance_dic[7573] = {'Paris','Key West'}
distance_dic [7729] ={'Paris','Havana'}
distance_dic[845] = {'Paris','Venice'}
distance_dic[6149] = {'Paris','Entebbe'}
distance_dic[5844] = {'Paris','New York'}
distance_dic[8034] = {'Paris','Ketchum'}

distance_dic [6644] ={'Milano','Toronto'}
distance_dic[7301] = {'Milano','Chicago'}
distance_dic[8173] = {'Milano','Key West'}
distance_dic[8326] ={'Milano','Havana'}
distance_dic [245] ={'Milano','Venice'}
distance_dic[5546] = {'Milano','Entebbe'}
distance_dic[6471] = {'Milano','New York'}
distance_dic[8657] = {'Milano','Ketchum'}

distance_dic [703] ={'Toronto','Chicago'}
distance_dic [2142] ={'Toronto','Key West'}
distance_dic [2306] ={'Toronto','Havana',}
distance_dic [6849] ={'Toronto','Venice'}
distance_dic [11752] ={'Toronto','Entebbe'}
distance_dic [556] ={'Toronto','New York'}
distance_dic [2792] ={'Toronto','Ketchum'}

distance_dic [2000] ={'Chicago','Key West'}
distance_dic [2140] ={'Chicago','Havana'}
distance_dic [7499] ={'Chicago','Venice'}
distance_dic [12455] ={'Chicago','Entebbe'}
distance_dic [1147] ={'Chicago','New York'}
distance_dic [2183] ={'Chicago','Ketchum'}

distance_dic [170] ={'Key West','Havana'}
distance_dic [8405] ={'Key West','Venice'}
distance_dic [12455] ={'Key West','Entebbe'}
distance_dic [1939] ={'Key West','New York'}
distance_dic [3643] ={'Key West','Ketchum'}

distance_dic [8559] ={'Havana','Venice'}
distance_dic [12544] ={'Havana','Entebbe'}
distance_dic [2109] ={'Havana','New York'}
distance_dic [3711] ={'Havana','Ketchum'}


distance_dic [5424] ={'Venice','Entebbe'}
distance_dic [6688] ={'Venice','New York'}
distance_dic [8807] ={'Venice','Ketchum'}

distance_dic [11394] ={'Entebbe','New York'}
distance_dic [14159] ={'Entebbe','Ketchum'}

distance_dic [3312] ={'New York','Ketchum'}

#Randomly choose a start city to begin 

In [49]:
start_city = (choices(city_list))[0]
print(f"Let's start at {start_city}!" )

Let's start at Paris!


#Define a list to store the route and a list to store the distance (shortest) in each step.

In [50]:
route = []

route_dist = []

##Use use greedy algorithm by find the shortest distance in every step.

In [51]:
for _ in range(len(city_list)-1):
    distance_list = []
    for distance, cities in distance_dic.items():
        if start_city in cities:
            distance_list.append(distance)
            
    shortest_dist = min(distance_list)
    route_dist.append(shortest_dist)
    nearest_cities = list(distance_dic[shortest_dist])
    
    nearest_cities. remove(start_city)
    arrive_city = nearest_cities[0]
    route.append([start_city,arrive_city])
    start_city = arrive_city
    for i in distance_list:
        del(distance_dic[i])


#Calculate the total distance.

In [52]:
total_distance = sum (route_dist)
print(f'Total distance is {total_distance}'km)

Total distance is 24862


#Print the route.

In [53]:
print('My route is:')
for j in range(len(route)):
    print(f'\t{j+1}.{route[j]}')
    

My route is:
	1.['Paris', 'Milano']
	2.['Milano', 'Venice']
	3.['Venice', 'Entebbe']
	4.['Entebbe', 'New York']
	5.['New York', 'Toronto']
	6.['Toronto', 'Chicago']
	7.['Chicago', 'Oak Park（Illinois）']
	8.['Oak Park（Illinois）', 'Key West']
	9.['Key West', 'Havana']
	10.['Havana', 'Ketchum']


# If I want to choose start city ... 

In [57]:



city_list = ['Oak Park（Illinois）','Paris','Milano','Toronto','Chicago','Key West','Havana',
               'Venice','Entebbe','New York','Ketchum']
print (city_list)

 #Choose a start city.
start_city1 = input('Choose a start city:\n')
print(f"Let's start at {start_city1}!" )
def greedy_travel(start_city):   
    distance_dic={}
    distance_dic[6667] = {'Oak Park（Illinois）','Paris'}
    distance_dic[7307] = {'Milano','Oak Park（Illinois）'}
    distance_dic[712] = {'Oak Park（Illinois）','Toronto'}
    distance_dic [12] ={'Oak Park（Illinois）','Chicago'}
    distance_dic[2007] = {'Oak Park（Illinois）','Key West'}
    distance_dic[2147] = {'Oak Park（Illinois）','Havana'}
    distance_dic[7505] = {'Oak Park（Illinois）','Venice'}
    distance_dic[12464] = {'Oak Park（Illinois）','Entebbe'}
    distance_dic[1158] = {'Oak Park（Illinois）','New York'}
    distance_dic[2171] = {'Oak Park（Illinois）','Ketchum'}
    
    distance_dic [640] ={'Paris','Milano'}
    distance_dic[6006] = {'Paris','Toronto'}
    distance_dic[6661] = {'Paris','Chicago'}
    distance_dic[7573] = {'Paris','Key West'}
    distance_dic [7729] ={'Paris','Havana'}
    distance_dic[845] = {'Paris','Venice'}
    distance_dic[6149] = {'Paris','Entebbe'}
    distance_dic[5844] = {'Paris','New York'}
    distance_dic[8034] = {'Paris','Ketchum'}
    
    distance_dic [6644] ={'Milano','Toronto'}
    distance_dic[7301] = {'Milano','Chicago'}
    distance_dic[8173] = {'Milano','Key West'}
    distance_dic[8326] ={'Milano','Havana'}
    distance_dic [245] ={'Milano','Venice'}
    distance_dic[5546] = {'Milano','Entebbe'}
    distance_dic[6471] = {'Milano','New York'}
    distance_dic[8657] = {'Milano','Ketchum'}
    
    distance_dic [703] ={'Toronto','Chicago'}
    distance_dic [2142] ={'Toronto','Key West'}
    distance_dic [2306] ={'Toronto','Havana',}
    distance_dic [6849] ={'Toronto','Venice'}
    distance_dic [11752] ={'Toronto','Entebbe'}
    distance_dic [556] ={'Toronto','New York'}
    distance_dic [2792] ={'Toronto','Ketchum'}
    
    distance_dic [2000] ={'Chicago','Key West'}
    distance_dic [2140] ={'Chicago','Havana'}
    distance_dic [7499] ={'Chicago','Venice'}
    distance_dic [12455] ={'Chicago','Entebbe'}
    distance_dic [1147] ={'Chicago','New York'}
    distance_dic [2183] ={'Chicago','Ketchum'}
    
    distance_dic [170] ={'Key West','Havana'}
    distance_dic [8405] ={'Key West','Venice'}
    distance_dic [12455] ={'Key West','Entebbe'}
    distance_dic [1939] ={'Key West','New York'}
    distance_dic [3643] ={'Key West','Ketchum'}
    
    distance_dic [8559] ={'Havana','Venice'}
    distance_dic [12544] ={'Havana','Entebbe'}
    distance_dic [2109] ={'Havana','New York'}
    distance_dic [3711] ={'Havana','Ketchum'}
    
    
    distance_dic [5424] ={'Venice','Entebbe'}
    distance_dic [6688] ={'Venice','New York'}
    distance_dic [8807] ={'Venice','Ketchum'}
    
    distance_dic [11394] ={'Entebbe','New York'}
    distance_dic [14159] ={'Entebbe','Ketchum'}
    
    distance_dic [3312] ={'New York','Ketchum'}
       
    #Define a list to store  the route
    route = []
    #Define a list to store the distance (shortest) in each step.
    route_dist = []
    #Use use greedy algorithm by find the shortest distance in every step.
    for _ in range(len(city_list)-1):
        distance_list = []
        for distance, cities in distance_dic.items():
            if start_city in cities:
                distance_list.append(distance)
       
                
        shortest_dist = min(distance_list)
        route_dist.append(shortest_dist)
        nearest_cities = list(distance_dic[shortest_dist])
        
        nearest_cities. remove(start_city)
        arrive_city = nearest_cities[0]
        route.append([start_city,arrive_city])
        start_city = arrive_city
        for i in distance_list:
            del(distance_dic[i])
    
    
    #Calculate the total distance.
    total_distance = sum (route_dist)
    print(f'Total distance is {total_distance}km')
    #Print the route.
    print('My route is:')
    for j in range(len(route)):
        print(f'\t{j+1}.{route[j]}')


greedy_travel(start_city1)

['Oak Park（Illinois）', 'Paris', 'Milano', 'Toronto', 'Chicago', 'Key West', 'Havana', 'Venice', 'Entebbe', 'New York', 'Ketchum']
Choose a start city:
Paris
Let's start at Paris!
Total distance is 24862km
My route is:
	1.['Paris', 'Milano']
	2.['Milano', 'Venice']
	3.['Venice', 'Entebbe']
	4.['Entebbe', 'New York']
	5.['New York', 'Toronto']
	6.['Toronto', 'Chicago']
	7.['Chicago', 'Oak Park（Illinois）']
	8.['Oak Park（Illinois）', 'Key West']
	9.['Key West', 'Havana']
	10.['Havana', 'Ketchum']


# Campare the result of each cities as start city with others and find the best solution and shortest distance.

In [56]:


city_list = ['Oak Park（Illinois）','Paris','Milano','Toronto','Chicago','Key West','Havana',
               'Venice','Entebbe','New York','Ketchum']


def greedy_shortestdist(start_city):   
    distance_dic={}
    distance_dic[6667] = {'Oak Park（Illinois）','Paris'}
    distance_dic[7307] = {'Milano','Oak Park（Illinois）'}
    distance_dic[712] = {'Oak Park（Illinois）','Toronto'}
    distance_dic [12] ={'Oak Park（Illinois）','Chicago'}
    distance_dic[2007] = {'Oak Park（Illinois）','Key West'}
    distance_dic[2147] = {'Oak Park（Illinois）','Havana'}
    distance_dic[7505] = {'Oak Park（Illinois）','Venice'}
    distance_dic[12464] = {'Oak Park（Illinois）','Entebbe'}
    distance_dic[1158] = {'Oak Park（Illinois）','New York'}
    distance_dic[2171] = {'Oak Park（Illinois）','Ketchum'}
    
    distance_dic [640] ={'Paris','Milano'}
    distance_dic[6006] = {'Paris','Toronto'}
    distance_dic[6661] = {'Paris','Chicago'}
    distance_dic[7573] = {'Paris','Key West'}
    distance_dic [7729] ={'Paris','Havana'}
    distance_dic[845] = {'Paris','Venice'}
    distance_dic[6149] = {'Paris','Entebbe'}
    distance_dic[5844] = {'Paris','New York'}
    distance_dic[8034] = {'Paris','Ketchum'}
    
    distance_dic [6644] ={'Milano','Toronto'}
    distance_dic[7301] = {'Milano','Chicago'}
    distance_dic[8173] = {'Milano','Key West'}
    distance_dic[8326] ={'Milano','Havana'}
    distance_dic [245] ={'Milano','Venice'}
    distance_dic[5546] = {'Milano','Entebbe'}
    distance_dic[6471] = {'Milano','New York'}
    distance_dic[8657] = {'Milano','Ketchum'}
    
    distance_dic [703] ={'Toronto','Chicago'}
    distance_dic [2142] ={'Toronto','Key West'}
    distance_dic [2306] ={'Toronto','Havana',}
    distance_dic [6849] ={'Toronto','Venice'}
    distance_dic [11752] ={'Toronto','Entebbe'}
    distance_dic [556] ={'Toronto','New York'}
    distance_dic [2792] ={'Toronto','Ketchum'}
    
    distance_dic [2000] ={'Chicago','Key West'}
    distance_dic [2140] ={'Chicago','Havana'}
    distance_dic [7499] ={'Chicago','Venice'}
    distance_dic [12455] ={'Chicago','Entebbe'}
    distance_dic [1147] ={'Chicago','New York'}
    distance_dic [2183] ={'Chicago','Ketchum'}
    
    distance_dic [170] ={'Key West','Havana'}
    distance_dic [8405] ={'Key West','Venice'}
    distance_dic [12455] ={'Key West','Entebbe'}
    distance_dic [1939] ={'Key West','New York'}
    distance_dic [3643] ={'Key West','Ketchum'}
    
    distance_dic [8559] ={'Havana','Venice'}
    distance_dic [12544] ={'Havana','Entebbe'}
    distance_dic [2109] ={'Havana','New York'}
    distance_dic [3711] ={'Havana','Ketchum'}
    
    
    distance_dic [5424] ={'Venice','Entebbe'}
    distance_dic [6688] ={'Venice','New York'}
    distance_dic [8807] ={'Venice','Ketchum'}
    
    distance_dic [11394] ={'Entebbe','New York'}
    distance_dic [14159] ={'Entebbe','Ketchum'}
    
    distance_dic [3312] ={'New York','Ketchum'}
       
    #Define a list to store  the route
    route = []
    #Define a list to store the distance (shortest) in each step.
    route_dist = []
    #Use use greedy algorithm by find the shortest distance in every step.
    for _ in range(len(city_list)-1):
        distance_list = []
        for distance, cities in distance_dic.items():
            if start_city in cities:
                distance_list.append(distance)
       
                
        shortest_dist = min(distance_list)
        route_dist.append(shortest_dist)
        nearest_cities = list(distance_dic[shortest_dist])
        
        nearest_cities. remove(start_city)
        arrive_city = nearest_cities[0]
        route.append([start_city,arrive_city])
        start_city = arrive_city
        for i in distance_list:
            del(distance_dic[i])
    
    
    #Calculate the total distance.
    total_distance = sum (route_dist)
 
    return total_distance 

startcities_shortestdict = {}
for name in city_list:
    startcities_shortestdict[name] = greedy_shortestdist(name)

for n,dist in startcities_shortestdict.items():
    if dist == min(startcities_shortestdict.values()):
        print(f'{n} as a start city is the best solution and the shortest distance is {dist} km.')

Entebbe as a start city is the best solution and the shortest distance is 19312 km.


#Conclusion


Greedy algorithm may not be the best way to solve TSP problem but I truly get my own solution. Entebbe as a start city is the best solution and the shortest distance is 19312 km.